vp/workflow/views.py

```diff
@@ -1,64 +1,73 @@
 from django.http import JsonResponse
+from .models import Workflow, WorkflowException
 import networkx as nx
 import json
 import os
 
 # Build paths inside the project like this: os.path.join(BASE_DIR, ...)
 BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
 
+# Initialize a Workflow object
+workflow = Workflow()
```

Comment:
I'm pretty sure this cannot be global, or every session on the server will be manipulating the same workflow instance.

Reply:
I think this is correct, and I had the same concern as well, but it was the easiest way I could think of to demonstrate some of the separation between modules.

> Every view function would then look up the session's workflow instance (by filename), load it, and operate on it. 

That's kind of where I ended up going in the `node/views.py` route. The first line is `workflow.retrieve_from_session(request)` and the last line is the call to `workflow.store_in_session(request)`, which works as you describe, albeit using the global `Workflow` instance which is not good. The problem I've run into with the models is an instance needs to be JSON-serializable in order to be stored in the session, which I haven't found an easy way around. Regular Django Models get around this by being loaded at startup/interacting with the database (not 100% sure)?

One possible solution might be to [write some middleware](https://docs.djangoproject.com/en/3.0/topics/http/middleware/) to imitate this a bit. On each request, it can create the Workflow object, and load the elements into the session (without the need for the global variable) before continuing on to the rest of the view logic.

This middleware approach may/may not be compatible with your suggestion on packaging the app(s). I definitely like this idea for the CLI requirement and we should definitely investigate further. I don't have any prior experience with that, so I'll defer to @diegostruk1 and @matthew-t-smith for now to hear their thoughts.

Reply:
Could we just store the filename in the session? Then initialize the `Workflow` instance within the handler?

Reply:
Love the middleware idea, and it's absolutely compatible with factoring out the workflow into its own package!

- `workflow` is its own package that defines the `Workflow` and `Node` classes currently in the Django app. Instances can be serialized by pickling, or we can define our own JSON serialization.
- Only the filename of the serialized workflow is stored in the session.
- A middleware class intercepts every request, looks up the filename from the session, and calls `Workflow.deserialize()` to create an instance and pass it to the view function (I think it's possible to store Python objects on the request, but not positive).
- The view function operates on the graph, and writes it back to disk (maybe that could be done by middleware in the response cycle)
- To run a workflow via CLI, the user provides the serialized workflow file (which they presumably saved from the server), and the `workflow` package loads and executes it, with no need for the Django application.

Again, just my thoughts. Definitely want to hear from the back-end guys.

Reply:
Following up with this discussion:
- Workflow package that includes both Node and Workflow classes: I think this makes perfect sense. As Samir mentions, this will be helpful for serializing. I think the both classes look great how they are now to start with. It will also improve performance of the CLI.
- Filename of the serialized workflow sounds good. I would like to test that a bit just to make sure that is really all we need. 
- Middleware class interceptor sounds great. I don't have much experience with doing this in  Django but I've done it before and seems like that link that Matt Thomas provided gives a good amount of information.